cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -395,42 +429,56 @@ private void deleteSchedule(Id.Program programId, SchedulableProgramType program
   @VisibleForTesting
   void registerTemplates() {
     try {
+      // generate a completely new map in case some templates were removed
+      Map<String, ApplicationTemplateInfo> newInfoMap = Maps.newHashMap();
+      Map<String, ApplicationTemplateInfo> newFileTemplateMap = Maps.newHashMap();
+
       File baseDir = new File(configuration.get(Constants.AppFabric.APP_TEMPLATE_DIR));
       Collection<File> files = FileUtils.listFiles(baseDir, new String[]{"jar"}, true);
       for (File file : files) {
-        try {
-          Manifest manifest = new JarFile(file.getAbsolutePath()).getManifest();
-          ApplicationTemplateInfo applicationTemplateInfo = createAppTemplateInfo(file, manifest);
-          if (applicationTemplateInfo != null) {
-            appTemplateInfos.put(applicationTemplateInfo.getName(), applicationTemplateInfo);
-          } else {
-            LOG.warn("Missing required information to create adapter {}", file.getAbsolutePath());
-          }
-        } catch (IOException e) {
-          LOG.warn(String.format("Unable to read adapter jar %s", file.getAbsolutePath()));
-        }
+        ApplicationTemplateInfo info = getTemplateInfo(file);
+        newInfoMap.put(info.getName(), info);
+        newFileTemplateMap.put(info.getFile().getName(), info);
       }
+      appTemplateInfos = newInfoMap;
+      fileToTemplateMap = newFileTemplateMap;
     } catch (Exception e) {
       LOG.warn("Unable to read the plugins directory");
     }
   }
 
-  // TODO: call configure on the application itself to get name and program type instead of requiring
-  //       them in the manifest
-  private ApplicationTemplateInfo createAppTemplateInfo(File file, Manifest manifest) {
-    if (manifest != null) {
-      Attributes mainAttributes = manifest.getMainAttributes();
+  private ApplicationTemplateInfo getTemplateInfo(File jarFile)
+    throws InterruptedException, ExecutionException, TimeoutException, IOException {
+    ApplicationTemplateInfo existing = fileToTemplateMap.get(jarFile.getAbsolutePath());
+    HashCode fileHash = Files.hash(jarFile, Hashing.md5());
+    // if the file is the same, just return
+    if (existing != null && fileHash.equals(existing.getFileHash())) {
+      return existing;
+    }
 
-      String name = mainAttributes.getValue(ApplicationTemplateManifestAttributes.NAME);
-      String programType = mainAttributes.getValue(ApplicationTemplateManifestAttributes.PROGRAM_TYPE);
-      return new ApplicationTemplateInfo(file, name, ProgramType.valueOf(programType.toUpperCase()));
+    // instantiate the template application and call configure() on it to determine it's specification
+    InMemoryConfigurator configurator = new InMemoryConfigurator(
```

Comment:
It is a bit confusing to see System namespace here. If we are deploying to usernamespace, perhaps add a comment or remove the variable in constructor.